sys/net/network_layer/ng_ipv6/ng_ipv6.c

```diff
@@ -0,0 +1,501 @@
+/*
+ * Copyright (C) 2015 Martine Lenders <mlenders@inf.fu-berlin.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ */
+#include <errno.h>
+#include <inttypes.h>
+#include <stdbool.h>
+
+#include "byteorder.h"
+#include "cpu-conf.h"
+#include "kernel_types.h"
+#include "net/ng_netbase.h"
+#include "net/ng_protnum.h"
+#include "thread.h"
+#include "utlist.h"
+
+#include "net/ng_ipv6/nc.h"
+#include "net/ng_ipv6/netif.h"
+
+#include "net/ng_ipv6.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define _MAX_L2_ADDR_LEN    (8U)
+
+static char _ipv6_stack[NG_IPV6_STACK_SIZE];
+static kernel_pid_t _ipv6_pid = KERNEL_PID_UNDEF;
+
+#if ENABLE_DEBUG
+static char addr_str[NG_IPV6_ADDR_MAX_STR_LEN];
+#endif
+
+static void _ipv6_receive(kernel_pid_t iface, ng_pktsnip_t *pkt);
+static void _ipv6_send(ng_pktsnip_t *pkt, bool prep_hdr);
+static void *_ipv6_event_loop(void *args);
+static void _dispatch_ipv6_pkt(ng_nettype_t type, uint32_t demux_ctx,
+                               ng_pktsnip_t *pkt);
+kernel_pid_t ng_ipv6_init(void)
+{
+    if (_ipv6_pid == KERNEL_PID_UNDEF) {
+        _ipv6_pid = thread_create(_ipv6_stack, NG_IPV6_STACK_SIZE, NG_IPV6_PRIO,
+                                  CREATE_STACKTEST, _ipv6_event_loop, NULL,
+                                  "ipv6");
+    }
+
+    return _ipv6_pid;
+}
+
+void ng_ipv6_demux(kernel_pid_t iface, ng_pktsnip_t *pkt, ng_pktsnip_t *payload,
+                   uint8_t nh)
+{
+    int receiver_num;
+
+    payload->type = ng_nettype_from_protnum(nh);
+
+    /* TODO: add ICMPv6 and extension header handling */
+
+    receiver_num = ng_netreg_num(payload->type, NG_NETREG_DEMUX_CTX_ALL) +
+                   ng_netreg_num(NG_NETTYPE_IPV6, nh);
+
+    if (receiver_num == 0) {
+        DEBUG("udp: unable to forward packet as no one is interested in it\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+    ng_pktbuf_hold(pkt, receiver_num - 1);
+    /* IPv6 is not interested anymore so `- 1` */
+    _dispatch_ipv6_pkt(payload->type, NG_NETREG_DEMUX_CTX_ALL, pkt);
+    _dispatch_ipv6_pkt(NG_NETTYPE_IPV6, nh, pkt);
+}
+
+/* internal functions */
+static inline ng_pktsnip_t *_start_write(ng_pktsnip_t *pkt)
+{
+    ng_pktsnip_t *tmp = ng_pktbuf_start_write(pkt);
+
+    if (tmp == NULL) {
+        ng_pktbuf_release(pkt);
+    }
+
+    return tmp;
+}
+
+static void *_ipv6_event_loop(void *args)
+{
+    msg_t msg, reply, msg_q[NG_IPV6_MSG_QUEUE_SIZE];
```

Comment:
If you know that the queue is allocated on the stack... But you're right will open an issue.